### PR TITLE
Specify overload name in function schema

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17779 Specify overload name in function schema**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14293878/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17780 Allow registering same operator schema multiple times&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14316791/)

The FunctionSchema can now store an overload name and the parser knows how to parse it. Specify like this:

    my_func.overload1(arg1: Tensor) -> Tensor
    my_func.overload2(arg1: Tensor, arg2: Tensor) -> Tensor

Differential Revision: [D14293878](https://our.internmc.facebook.com/intern/diff/D14293878/)

### CODE REVIEW

**Comment by apaszke** [comment]
Sorry, what are overload names for? If you really need multiple functions of the same signature, why can't you just call them differently??

**Comment by smessmer** [comment]
This is for multiple functions with different signature but the same name. Some operators need overloads.

In a world with unnamed overloads, we still need a way to register overloads, so if a user registers a kernel against an operator with a mismatching signature we have to allow that because we don't know if (a) their kernel implementation has the wrong signature (which is an error) or (b) they actually want to register a new overload. This is bad UX since we can't warn about errors.
In a world with named overloads, users can register new overloads by giving them a different overload name. We can then check that the kernel signature must exactly match the schema for this overload and report errors if it mismatches.

Later on, named overloads could be used for other things too, but this is the main use case for now.

**Comment by apaszke** [comment]
So in that case you're saying that function name + overload name should uniquely identify a signature? To be honest I'm not sure if that's all that safer. Just like you can supply an incorrect signature to a function name where you wanted to add a new backend, you can simply have a typo in the overload name. What's the difference?

**Comment by smessmer** [comment]
Yes, function name + overload name uniquely identify a schema (i.e. signature + annotations + whatever else is part of the schema).
In case of a typo in the overload name, you are out of luck, same as in the case of a typo in the operator name. This is, however, less likely to happen than a typo in a long function signature.

Basically, overload names force kernel writers to explicitly specify when they're overloading something instead of having it happen automatically, and that allows us to capture user intent more reliably.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 1]
```diff
@@ -448,6 +448,7 @@ RegisterOperators reg(
      Operator(
          FunctionSchema(
              "aten::warn",
+             "",
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: Can you just replace this with:
"aten::warn(str message, *, int stacklevel=2)"

It must have been added before we supported the kwarg_only annotation.

**Review by zdevito — APPROVED** [review]
LGTM. 

Context: we are going to start registering multiple implementations of the same overload for different backends. We want to avoid situations like:

// file1
register("foo(int a, int b)",  CPU, <impl>)

// file2
register("foo(int a, int b)",  SUPER_RARE_HIDDEN_BACKEND, <impl>)

// person unaware of the definition for SUPER_RARE_HIDDEN_BACKEND then edits file1:
register("foo(int a, Tensor b)",  CPU, <impl>)

// now it all appears to work until the first time someone tries to call foo with SUPER_RARE_HIDDEN_BACKEND and gets an unimplemetd error.

By optionally allowing an overload name, we can catch this error on registration and report a signature mismatch rather than on use.

While changing the name of the function can result in a similar problem, I think changing operator names is treated with more scrutiny than editing the argument list to, e.g., make it more permissive.

File: caffe2/core/c10_operator.h [hunk 2]
```diff
@@ -88,8 +88,10 @@ inline c10::FunctionSchema make_function_schema_for_c10(const char* OperatorName
   }
 
   return c10::FunctionSchema(
-    std::string("_caffe2::") + OperatorName,
-    std::move(actual_inputs), std::move(outputs));
+      std::string("_caffe2::") + OperatorName,
+      "caffe2",
```
[the remark this answers is not in the document]
Comment (smessmer) [reply caffe2/core/c10_operator.h]: Ok, switching to empty string instead.

**Comment by smessmer** [comment]
Closing and re-exporting because of ghexport issues

File: torch/csrc/jit/register_prim_ops.cpp (hunk 1 above)
Comment (smessmer) [reply torch/csrc/jit/register_prim_ops.cpp]: hm that fails with https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-devtoolset7-rocmrpm-centos7.5-test/4357/console . Reverting and using just the name for now.

**Pull request closed without merging**